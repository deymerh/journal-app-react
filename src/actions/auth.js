import { firebase, gogleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

export const loginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(12345, 'Pedro'))
    }, 3500);
  };
};
export const startGoogleLogin = () => {
  return (dispath) => {
    firebase.auth().signInWithPopup(gogleAuthProvider)
      .then(({ user }) => {
        dispath(login(user.uid, user.displayName))
      })
  }
}
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});