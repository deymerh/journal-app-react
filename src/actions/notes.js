import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/journalAdd';
import { types } from '../types/types';

export const startNewNote = ()=>{
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
};

export const startLoadingNotes = (uid)=>{
  return async (dispatch)=>{
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const activeNote = (id, note)=>({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note)=>({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes)=>({
  type:types.notesLoad,
  payload: notes
});

export const startSaveNote = (note)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    dispatch(refreshNote(note));
    Swal.fire('Saved', note.title, 'success');
  }
};

export const refreshNote = (note)=>({
  type: types.notesUpdated,
  payload:note
});

export const startUploadImg = (image)=>{
  return async(dispatch, getState)=>{
    const {active:noteActive} = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: ()=>{
        Swal.showLoading();
      }
    });
    const { secure_url } = await fileUpload(image);
    noteActive.url = secure_url;
    dispatch(startSaveNote(noteActive));
    Swal.close();
  }
};

export const startDeleteNoteSelected = (id)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id)=>({
  type: types.noteSelectDelete,
  payload: id
});