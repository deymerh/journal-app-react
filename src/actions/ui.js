import { types } from "../types/types"

export const setErrorAction = (err) => ({
  type: types.uiSetError,
  payload: err
});

export const setRemoveErrorAction = () => ({
  type: types.uiRemoveError,
});