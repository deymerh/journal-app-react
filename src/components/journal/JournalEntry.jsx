import React from 'react';
import moment from 'moment';
import { activeNote } from '../../actions/notes';
import { useDispatch } from 'react-redux';

export const JournalEntry = ({id, date, title, body, url}) => {
  const noteDate = moment(date);
  const dispatch = useDispatch();
  const handleActivatedNote = ()=>{
    dispatch(activeNote(id, {date, title, body, url}));
  };
  return (
    <div className="journal__entry animate__animated animate__fadeIn animated__fater"
      onClick={handleActivatedNote}
      data-testid="selectNote"    
    >
      {
        (url)&&  
          <img
            className="journal__entry-picture"
            src={url}
            style={{
              backgroundSize: 'cover'
          }}
          alt="note"
        />
      }
      <div className="jornal__entry-body">
        <p className="jornal__entry-title">{title}</p>
        <p className="jornal__entry-content">{body}</p>
      </div>
      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  )
}
