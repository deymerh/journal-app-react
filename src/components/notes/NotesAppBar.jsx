import React, { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploadImg } from '../../actions/notes';

export const NotesAppBar = () => {
  const {active:noteActive} = useSelector(state => state.notes)
  const dispatch = useDispatch();
  const inputFileImg = useRef(null);
  const handleSaveNote = ()=>{
    dispatch(startSaveNote(noteActive));
  }
  const handleUploadPicture = () => {
    inputFileImg.current.click();
    console.log('Click');
  };
  const handleFileChange = (e)=>{
    const image = e.target.files[0];
    if (image) {
      dispatch(startUploadImg(image));
    }
  };
  return (
    <div className="notes__appbar">
      <span>28 de Agosto 2021</span>
      <input
        type="file"
        id="fileSelector"
        name="fileSelector"
        style={{display:'none'}}
        onChange={handleFileChange}
        ref={inputFileImg}
      />
      <div>
        <button
          className="btn"
          onClick={handleUploadPicture}
        >
          Picture
        </button>
        <button className="btn"
          onClick={handleSaveNote}
        >
          Save
        </button>
      </div>
    </div>
  )
}
