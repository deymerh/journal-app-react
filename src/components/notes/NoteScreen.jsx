import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleteNoteSelected } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
  const {active:noteActive} = useSelector(state => state.notes);
  const dispatch = useDispatch();
  const [values, handleInputChange, reset] = useForm(noteActive);
  const {title, body} =  values;
  const activeNotedId =  useRef(noteActive.id);
  useEffect(() => {
    if (noteActive.id !== activeNotedId.current) {
      reset(noteActive);
      activeNotedId.current = noteActive.id;
    }
  }, [noteActive, reset]);
  useEffect(() => {
    dispatch(activeNote(noteActive.id, {...values}))
  }, [values, dispatch]);
  const handleDelete = ()=>{
    dispatch(startDeleteNoteSelected(noteActive.id));
  };
  return (
    <div className="notes__main-content animate__animated animate__fadeIn animated__fater">
      <NotesAppBar />
      <div className="notes__content">
        <input
          name="title"
          value={title}
          type="text"
          id=""
          placeholder="Some awesome title"
          className="notes__title-input"
          onChange={handleInputChange}
        />
        <textarea
          name="body"
          value={body}
          id=""
          cols="30"
          rows="10"
          placeholder="what happened today"
          className="notes__textarea"
          onChange={handleInputChange}
        >
        </textarea>
        {
          (noteActive.url) &&
            (
              <div className="notes__image">
                <img src={noteActive.url} alt="Imagen" />
              </div>
            )
        }
      </div>
      <button
        className="btn btn-danger"
        onClick={handleDelete}
      >
        Delete
      </button>
    </div>
  )
}
