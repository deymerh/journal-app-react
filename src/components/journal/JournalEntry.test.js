import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { JournalEntry } from './JournalEntry';
import { activeNote } from '../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {}
let store = mockStore(initState); 

describe('Pruabas en el <JournalEntry />', () => {
  const note = {
    id: 123456,
    date: 0,
    title: 'Hola',
    body: 'mundo',
    url: 'https://www.deymerh.co.mifoto.jpg'
  };
  test('Renderizarse correctamente - `[snapshot]`', () => {
    store.dispatch = jest.fn();
    const {container} = render(
      <Provider store={store}>
        <JournalEntry {...note}/>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
  test('Debe activar la nota', async() => {
    store.dispatch = jest.fn();
    const {findByTestId} = render(
      <Provider store={store}>
        <JournalEntry {...note}/>
      </Provider>
    );
    const selectNote = await findByTestId('selectNote');
    fireEvent.click(selectNote);
    expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, {...note}));
  });
  
});