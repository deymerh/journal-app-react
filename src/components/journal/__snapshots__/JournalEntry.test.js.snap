// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruabas en el <JournalEntry /> Renderizarse correctamente - \`[snapshot]\` 1`] = `
<div>
  <div
    class="journal__entry animate__animated animate__fadeIn animated__fater"
    data-testid="selectNote"
  >
    <img
      alt="note"
      class="journal__entry-picture"
      src="https://www.deymerh.co.mifoto.jpg"
      style="background-size: cover;"
    />
    <div
      class="jornal__entry-body"
    >
      <p
        class="jornal__entry-title"
      >
        Hola
      </p>
      <p
        class="jornal__entry-content"
      >
        mundo
      </p>
    </div>
    <div
      class="journal__entry-date-box"
    >
      <span>
        Wednesday
      </span>
      <h4>
        31st
      </h4>
    </div>
  </div>
</div>
`;
